cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ReentrantDistributedLock.java

```diff
@@ -0,0 +1,395 @@
+          completion.setException(t);
+        }
+      }
+    });
+
+    // Gets the result from the completion
+    try {
+      if (interruptible) {
+        localLockNode.set(completion.get(timeout, unit));
+      } else {
+        localLockNode.set(Uninterruptibles.getUninterruptibly(completion, timeout, unit));
+      }
+      return true;
+    } catch (InterruptedException e) {
+      completion.cancel(true);
+      throw e;
+    } catch (TimeoutException e) {
+      completion.cancel(true);
+      throw e;
+    } catch (CancellationException e) {
+      // If the completion get cancelled, meaning the lock acquisition is aborted.
+      return false;
+    }
+  }
+
+  /**
+   * Executes the lock acquisition process. This corresponds to step 2-5 of the distributed lock receipt.
+   */
+  private void doAcquire(final SettableFuture<String> completion, final boolean waitForLock,
+                         final String guid, @Nullable final String lockPath) {
+    // Step 2. Get all children under the lock parent.
+    Futures.addCallback(zkClient.getChildren(path), new FutureCallback<NodeChildren>() {
+
+      @Override
+      public void onSuccess(NodeChildren children) {
+        // Find the lock node in case the creation step failed by matching the guid
+        // See "Recoverable Errors and the GUID" in the ZooKeeper guide
+        final String lockNode = lockPath == null ? findLockNode(children.getChildren(), guid) : lockPath;
+        if (lockNode == null) {
+          // If not able to find the lock node, fail the locking procedure.
+          completion.setException(new IllegalStateException("Failed to acquire lock").fillInStackTrace());
+          return;
+        }
+
+        if (lockPath == null) {
+          // If lock node was not determined in step 1 due to connection loss exception, need to add the
+          // node deletion handler in here after the actual lockNode is determined.
+          deleteNodeOnFailure(completion, lockNode);
+        }
+
+        // Find the node to watch, which is the one with the largest id that is smaller than currentId
+        // If the current id is the smallest one, nodeToWatch will be null
+        String nodeToWatch = findNodeToWatch(children, lockNode, guid);
+
+        // Step 3a. lockNode is the lowest, hence this become lock owner.
+        if (nodeToWatch == null) {
+          // Acquired lock
+          completion.set(lockNode);
+        } else if (!waitForLock) {
+          // This is for the case of tryLock() without timeout.
+          completion.cancel(true);
+        }
+        // If the lock acquisition is completed, due to whatever reason, we need to watch for any other nodes
+        if (completion.isDone()) {
+          return;
+        }
+
+        // Step 3b and 4. Call exists on the next lowest sequence ID
+        OperationFuture<Stat> existsFuture = zkClient.exists(nodeToWatch, new Watcher() {
+          @Override
+          public void process(WatchedEvent event) {
```

Comment:
I am not sure. ZooKeeper guarantees well-ordered sequence numbers. But does it also guarantee that the clients' calls return in the same order, and that they become visible to all clients in that order? I could see this scenario (although I don't know how exactly ZK implements sequencing):

- Assume client 1 wins, client 2 is second, client 3 third.
- There is a race between clients 2 and 3. Although client 2 received a smaller timestamp, client 3's create call returns before node #2 is visible. So the listChildren of client 3 does not show node #2, and therefore client 3 watches node #1. 
- Client 2 also watches node #1. 
- When node #1 goes away, both clients 2 and 3 wake up. Only one of them can win, so they both have to list children again and determine whether they are first. 

Could that happen? I am wondering... but if it could happen, then doesn't that mean that the winner cannot be sure of winning? Because if he is not guaranteed to see all nodes smaller than his own, how can he know that he has the smallest?

So perhaps you are right. But repeating step 2 is certainly safer.